File: integrations/unity/com.lazydeck.editor/Editor/Api/LazyDeckClient.cs
using System;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

namespace LazyDeck.Editor.Api
{
    [Serializable]
    internal sealed class ApiErrorPayload
    {
        public string kind;
        public string message;
    }

    [Serializable]
    internal sealed class ApiErrorEnvelope
    {
        public ApiErrorPayload error;
    }

    /// <summary>
    /// The outcome of one LazyDeckClient request: either Ok with the raw
    /// JSON response body (callers deserialize with JsonUtility themselves
    /// into whatever shape that endpoint returns, since response shapes
    /// vary per route), or a structured error with a stable Kind, mirroring
    /// the API's own {"error": {"kind", "message"}} envelope (see
    /// api/openapi.yaml's ApiErrorEnvelope).
    /// </summary>
    internal readonly struct ApiResult
    {
        public readonly bool Ok;
        public readonly long Status;
        public readonly string Body;
        public readonly string ErrorKind;
        public readonly string ErrorMessage;

        private ApiResult(bool ok, long status, string body, string errorKind, string errorMessage)
        {
            Ok = ok;
            Status = status;
            Body = body;
            ErrorKind = errorKind;
            ErrorMessage = errorMessage;
        }

        public static ApiResult Success(long status, string body) =>
            new ApiResult(true, status, body, null, null);

        public static ApiResult Failure(long status, string kind, string message) =>
            new ApiResult(false, status, null, kind, message);
    }

    /// <summary>
    /// Thin HTTP client for the lazydeck local service API (see #13 /
    /// api/openapi.yaml), the Unity counterpart of the Godot plugin's
    /// api/client.gd (#14/#16). Every method awaits its UnityWebRequest and
    /// returns an ApiResult rather than typed response classes per
    /// endpoint, since callers here are just editor UI code reading a
    /// couple of fields out of the JSON body.
    /// </summary>
    internal sealed class LazyDeckClient
    {
        // UnityWebRequest.timeout defaults to 0 (no timeout); every request
        // this client makes is a short, non-streaming call (connect,
        // devices, pair, discover), so a fixed bound is enough to guarantee
        // RequestAsync always completes instead of leaving a caller's
        // _busy flag stuck true if serve accepts a request and then stops
        // responding.
        private const int RequestTimeoutSeconds = 30;

        private readonly string _baseUrl;
        private readonly string _token;

        public LazyDeckClient(ConnectionInfo connection)
        {
            _baseUrl = connection.base_url;
            _token = connection.token;
        }

        /// <summary>
        /// Performs one request and awaits its completion. path is a
        /// /v1/... route; jsonBody, when non-null, is sent as the request
        /// body with a JSON content type.
        ///
        /// Bridges UnityWebRequestAsyncOperation.completed into an
        /// Awaitable via AwaitableCompletionSource, rather than polling
        /// with `while (!operation.isDone) await Awaitable.NextFrameAsync();`
        /// — that poll is tied to the player-frame loop, which an
        /// EditorWindow invokes outside Play mode only on its own repaint
        /// cadence, so it could leave a request pending indefinitely in an
        /// editor-only context. This class was written without a Unity
        /// Editor available to verify either approach against (see
        /// integrations/unity/README.md); this is the more
        /// editor-context-safe of the two per Unity's own Awaitable docs.
        /// </summary>
        public async Awaitable<ApiResult> RequestAsync(string method, string path, string jsonBody = null)
        {
            using var request = new UnityWebRequest(_baseUrl + path, method);
            request.SetRequestHeader("Authorization", "Bearer " + _token);
            request.downloadHandler = new DownloadHandlerBuffer();
            request.timeout = RequestTimeoutSeconds;
            if (!string.IsNullOrEmpty(jsonBody))
            {
                request.uploadHandler = new UploadHandlerRaw(Encoding.UTF8.GetBytes(jsonBody));
                request.SetRequestHeader("Content-Type", "application/json");
            }

            UnityWebRequestAsyncOperation operation = request.SendWebRequest();
            var completionSource = new AwaitableCompletionSource();
            operation.completed += _ => completionSource.SetResult();
            await completionSource.Awaitable;

            if (
                request.result == UnityWebRequest.Result.ConnectionError
                || request.result == UnityWebRequest.Result.DataProcessingError
            )
            {
                return ApiResult.Failure(0, "unreachable", $"request did not complete: {request.error}");
            }

            long status = request.responseCode;
            string body = request.downloadHandler.text;

            if (status >= 200 && status < 300)
            {
                return ApiResult.Success(status, body);
            }

            string kind = "unknown";
            string message = $"request failed with status {status}";
            if (!string.IsNullOrEmpty(body))
            {
                try
                {
                    ApiErrorEnvelope envelope = JsonUtility.FromJson<ApiErrorEnvelope>(body);
                    if (envelope?.error != null)
                    {
                        kind = envelope.error.kind;
                        message = envelope.error.message;
                    }
                }
                catch (ArgumentException)
                {
                    // Body wasn't a parseable error envelope; keep the generic message.
                }
            }
            return ApiResult.Failure(status, kind, message);
        }

        public Awaitable<ApiResult> GetHealthAsync() => RequestAsync("GET", "/v1/health");

        public Awaitable<ApiResult> GetCapabilitiesAsync() => RequestAsync("GET", "/v1/capabilities");

        public Awaitable<ApiResult> ListDevicesAsync() => RequestAsync("GET", "/v1/devices");

        public Awaitable<ApiResult> DiscoverDevicesAsync(float timeoutSeconds = 5f)
        {
            string body = JsonUtility.ToJson(new DiscoverRequest { timeout_seconds = timeoutSeconds });
            return RequestAsync("POST", "/v1/devices/discover", body);
        }

        public Awaitable<ApiResult> PairDeviceAsync(string deviceId) =>
            RequestAsync("POST", $"/v1/devices/{UnityWebRequest.EscapeURL(deviceId)}/pair");

        public Awaitable<ApiResult> GetDeviceStatusAsync(string deviceId) =>
            RequestAsync("GET", $"/v1/devices/{UnityWebRequest.EscapeURL(deviceId)}/status");

        /// <summary>
        /// Submits a deploy job for deviceId. directory must be an absolute
        /// path on this workstation (the API rejects a relative one — see
        /// api/openapi.yaml's deployments endpoint). Returns immediately with
        /// the queued job's snapshot; poll GetJobAsync to observe progress.
        /// </summary>
        public Awaitable<ApiResult> SubmitDeploymentAsync(
            string deviceId,
            string gameId,
            string directory,
            bool deleteExtraneous = false
        )
        {
            string body = JsonUtility.ToJson(
                new DeploymentRequest
                {
                    game_id = gameId,
                    directory = directory,
                    delete_extraneous = deleteExtraneous,
                }
            );
            return RequestAsync(
                "POST",
                $"/v1/devices/{UnityWebRequest.EscapeURL(deviceId)}/deployments",
                body
            );
        }

        /// <summary>
        /// Submits a log-sync job for deviceId. gameId is accepted for forward
        /// compatibility but currently unused by the backend (it always fetches
        /// the device's complete Steam logs/minidumps) — see api/openapi.yaml.
        ///
        /// Two request DTOs rather than one nullable field: JsonUtility always
        /// serializes every public field, so an empty gameId on a shared DTO
        /// would send `"game_id": ""` instead of omitting the key, which is a
        /// different request than the Godot client sends.
        /// </summary>
        public Awaitable<ApiResult> SubmitLogsSyncAsync(
            string deviceId,
            string directory,
            string gameId = ""
        )
        {
            string body = string.IsNullOrEmpty(gameId)
                ? JsonUtility.ToJson(new LogsSyncRequest { directory = directory })
                : JsonUtility.ToJson(
                    new LogsSyncRequestWithGame { directory = directory, game_id = gameId }
                );
            return RequestAsync(
                "POST",
                $"/v1/devices/{UnityWebRequest.EscapeURL(deviceId)}/logs/sync",
                body
            );
        }

        public Awaitable<ApiResult> GetJobAsync(string jobId) =>
            RequestAsync("GET", $"/v1/jobs/{UnityWebRequest.EscapeURL(jobId)}");

        public Awaitable<ApiResult> CancelJobAsync(string jobId) =>
            RequestAsync("DELETE", $"/v1/jobs/{UnityWebRequest.EscapeURL(jobId)}");

        [Serializable]
        private sealed class DiscoverRequest
        {
            public float timeout_seconds;
        }

        [Serializable]
        private sealed class DeploymentRequest
        {
            public string game_id;
            public string directory;
            public bool delete_extraneous;
        }

        [Serializable]
        private sealed class LogsSyncRequest
        {
            public string directory;
        }

        [Serializable]
        private sealed class LogsSyncRequestWithGame
        {
            public string directory;
            public string game_id;
        }
    }
}

File: integrations/unreal/LazyDeck/Source/LazyDeckEditor/Private/LazyDeckClient.cpp
#include "LazyDeckClient.h"

#include "Dom/JsonObject.h"
#include "GenericPlatform/GenericPlatformHttp.h"
#include "HttpModule.h"
#include "Interfaces/IHttpRequest.h"
#include "Interfaces/IHttpResponse.h"
#include "Serialization/JsonReader.h"
#include "Serialization/JsonSerializer.h"
#include "Serialization/JsonWriter.h"

// Every request this client makes is a short, non-streaming call (connect,
// devices, pair, discover); a fixed bound guarantees the request delegate
// always fires instead of leaving a caller's busy flag stuck true if serve
// accepts a request and then stops responding. Matches the Unity client's
// RequestTimeoutSeconds.
static constexpr float LazyDeckRequestTimeoutSeconds = 30.0f;

FLazyDeckClient::FLazyDeckClient(const FLazyDeckConnectionInfo& InConnection) : BaseUrl(InConnection.BaseUrl), Token(InConnection.Token) {}

void FLazyDeckClient::Request(const FString& Verb, const FString& Path, const FString& JsonBody, FLazyDeckApiResultDelegate OnComplete) const
{
	const TSharedRef<IHttpRequest> HttpRequest = FHttpModule::Get().CreateRequest();
	HttpRequest->SetURL(BaseUrl + Path);
	HttpRequest->SetVerb(Verb);
	HttpRequest->SetHeader(TEXT("Authorization"), TEXT("Bearer ") + Token);
	HttpRequest->SetTimeout(LazyDeckRequestTimeoutSeconds);
	if (!JsonBody.IsEmpty())
	{
		HttpRequest->SetHeader(TEXT("Content-Type"), TEXT("application/json"));
		HttpRequest->SetContentAsString(JsonBody);
	}

	HttpRequest->OnProcessRequestComplete().BindLambda(
		[OnComplete](FHttpRequestPtr, FHttpResponsePtr Response, bool bConnectedSuccessfully)
		{
			FLazyDeckApiResult Result;
			if (!bConnectedSuccessfully || !Response.IsValid())
			{
				Result.bOk = false;
				Result.ErrorKind = TEXT("unreachable");
				Result.ErrorMessage = TEXT("request did not complete");
				OnComplete.ExecuteIfBound(Result);
				return;
			}

			Result.StatusCode = Response->GetResponseCode();
			Result.Body = Response->GetContentAsString();

			if (Result.StatusCode >= 200 && Result.StatusCode < 300)
			{
				Result.bOk = true;
				OnComplete.ExecuteIfBound(Result);
				return;
			}

			Result.bOk = false;
			Result.ErrorKind = TEXT("unknown");
			Result.ErrorMessage = FString::Printf(TEXT("request failed with status %d"), Result.StatusCode);

			TSharedPtr<FJsonObject> JsonObject;
			TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(Result.Body);
			if (FJsonSerializer::Deserialize(Reader, JsonObject) && JsonObject.IsValid())
			{
				const TSharedPtr<FJsonObject>* ErrorObject;
				if (JsonObject->TryGetObjectField(TEXT("error"), ErrorObject))
				{
					(*ErrorObject)->TryGetStringField(TEXT("kind"), Result.ErrorKind);
					(*ErrorObject)->TryGetStringField(TEXT("message"), Result.ErrorMessage);
				}
			}
			OnComplete.ExecuteIfBound(Result);
		});

	HttpRequest->ProcessRequest();
}

void FLazyDeckClient::GetHealth(FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("GET"), TEXT("/v1/health"), FString(), OnComplete);
}

void FLazyDeckClient::GetCapabilities(FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("GET"), TEXT("/v1/capabilities"), FString(), OnComplete);
}

void FLazyDeckClient::ListDevices(FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("GET"), TEXT("/v1/devices"), FString(), OnComplete);
}

void FLazyDeckClient::DiscoverDevices(float TimeoutSeconds, FLazyDeckApiResultDelegate OnComplete) const
{
	const TSharedRef<FJsonObject> JsonObject = MakeShared<FJsonObject>();
	JsonObject->SetNumberField(TEXT("timeout_seconds"), TimeoutSeconds);
	FString Body;
	const TSharedRef<TJsonWriter<>> Writer = TJsonWriterFactory<>::Create(&Body);
	FJsonSerializer::Serialize(JsonObject, Writer);
	Request(TEXT("POST"), TEXT("/v1/devices/discover"), Body, OnComplete);
}

void FLazyDeckClient::PairDevice(const FString& DeviceId, FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("POST"), FString::Printf(TEXT("/v1/devices/%s/pair"), *FGenericPlatformHttp::UrlEncode(DeviceId)), FString(), OnComplete);
}

void FLazyDeckClient::GetDeviceStatus(const FString& DeviceId, FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("GET"), FString::Printf(TEXT("/v1/devices/%s/status"), *FGenericPlatformHttp::UrlEncode(DeviceId)), FString(), OnComplete);
}

void FLazyDeckClient::ListGames(const FString& DeviceId, FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("GET"), FString::Printf(TEXT("/v1/devices/%s/games"), *FGenericPlatformHttp::UrlEncode(DeviceId)), FString(), OnComplete);
}

void FLazyDeckClient::SubmitDeployment(const FString& DeviceId, const FString& GameId, const FString& Directory, bool bDeleteExtraneous,
									   FLazyDeckApiResultDelegate OnComplete) const
{
	const TSharedRef<FJsonObject> JsonObject = MakeShared<FJsonObject>();
	JsonObject->SetStringField(TEXT("game_id"), GameId);
	JsonObject->SetStringField(TEXT("directory"), Directory);
	JsonObject->SetBoolField(TEXT("delete_extraneous"), bDeleteExtraneous);
	FString Body;
	const TSharedRef<TJsonWriter<>> Writer = TJsonWriterFactory<>::Create(&Body);
	FJsonSerializer::Serialize(JsonObject, Writer);
	Request(TEXT("POST"), FString::Printf(TEXT("/v1/devices/%s/deployments"), *FGenericPlatformHttp::UrlEncode(DeviceId)), Body, OnComplete);
}

void FLazyDeckClient::SyncLogs(const FString& DeviceId, const FString& Directory, const FString& GameId, FLazyDeckApiResultDelegate OnComplete) const
{
	const TSharedRef<FJsonObject> JsonObject = MakeShared<FJsonObject>();
	JsonObject->SetStringField(TEXT("directory"), Directory);
	if (!GameId.IsEmpty())
	{
		JsonObject->SetStringField(TEXT("game_id"), GameId);
	}
	FString Body;
	const TSharedRef<TJsonWriter<>> Writer = TJsonWriterFactory<>::Create(&Body);
	FJsonSerializer::Serialize(JsonObject, Writer);
	Request(TEXT("POST"), FString::Printf(TEXT("/v1/devices/%s/logs/sync"), *FGenericPlatformHttp::UrlEncode(DeviceId)), Body, OnComplete);
}

void FLazyDeckClient::GetJob(const FString& JobId, FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("GET"), FString::Printf(TEXT("/v1/jobs/%s"), *FGenericPlatformHttp::UrlEncode(JobId)), FString(), OnComplete);
}

void FLazyDeckClient::CancelJob(const FString& JobId, FLazyDeckApiResultDelegate OnComplete) const
{
	Request(TEXT("DELETE"), FString::Printf(TEXT("/v1/jobs/%s"), *FGenericPlatformHttp::UrlEncode(JobId)), FString(), OnComplete);
}
